tools/make_patch.c

```diff
@@ -0,0 +1,597 @@
+			buffer_index = 0;
+			value = -1;
+			while (c != '\n' && c != '\r' && c != EOF) {
+				c = getc(file);
+			}
+		}
+	}
+
+	fclose(file);
+	free_buffer(buffer);
+	return *symbols;
+}
+
+void interpret_command(
+	char *command, const struct Symbol *current_hook, const struct Symbol *symbols,
+	struct Patches *patches, FILE *new_rom, FILE *orig_rom, FILE *output
+) {
+	// Count the arguments
+	int argc = 0;
+	for (char *c = command; *c; c++) {
+		if (*c == ':') {
+			argc++;
+		}
+	}
+
+	// Get the arguments
+	char *argv[argc]; // VLA
+	char *arg = command;
+	for (int i = 0; i < argc; i++) {
+		arg = strchr(arg, ':');
+		if (!arg) {
+			break;
+		}
+		*arg++ = '\0';
+		argv[i] = arg;
+	}
+
+	int offset = current_hook ? current_hook->offset : -1;
+
+	// Use the arguments
+	if (!strcmp(command, "ADDREss")) {
+		// This is only necessary to match the exact upper/lower casing in the original patch
+		fprintf(output, "0x%X%x", HIGH(offset), LOW(offset));
+
+	} else if (!strcmp(command, "Address") || !strcmp(command, "address")) {
+		if (argc > 0) {
+			offset += strtol(argv[0], NULL, 0);
+		}
+		fprintf(output, "0x");
+		if (argc > 1) {
+			for (int i = strtol(argv[1], NULL, 0); i > 0; i--) {
+				putc('0', output);
+			}
+		}
+		fprintf(output, isupper((unsigned char)command[0]) ? "%X" : "%x", offset);
+
+	} else if (!strcmp(command, "Patch") || !strcmp(command, "patch")) {
+		if (argc > 0) {
+			offset += strtol(argv[0], NULL, 0);
+		}
+		if (fseek(orig_rom, offset, SEEK_SET)) {
+			fprintf(stderr, "Error: Could not seek to the offset of %s in the original ROM\n", current_hook->name);
+			return;
+		}
+		if (fseek(new_rom, offset, SEEK_SET)) {
+			fprintf(stderr, "Error: Could not seek to offset of %s in the new ROM\n", current_hook->name);
+			return;
+		}
+		if (argc <= 1 || strcmp(argv[1], "big")) {
+			int c = getc(new_rom);
+			if (c == getc(orig_rom)) {
+				fprintf(stderr, "Warning: %s doesn't actually contain any differences\n", current_hook->name);
+			}
+			append_patch(patches, offset, 1);
+			fprintf(output, "0x");
+			fprintf(output, isupper((unsigned char)command[0]) ? "%02X" : "%02x", c);
+		} else {
+			char *searchend = xmalloc(strlen(current_hook->name) + strlen("_End") + 1);
+			strcpy(searchend, current_hook->name);
+			strcat(searchend, "_End");
+			current_hook = find_symbol(symbols, searchend);
+			if (!current_hook) {
+				fprintf(stderr, "Error: Could not find symbol: %s", searchend);
+			} else {
+				// Figure out the length of the patch
+				int length = current_hook->offset - offset;
+				memset(searchend, 0, (strlen(current_hook->name)));
+				// We've got the length, now go back
+				fseek(new_rom, offset, SEEK_SET);
+				// Print out the patch
+				append_patch(patches, offset, length);
+				fprintf(output, "a%d:", length);
+				for (int i = 0; i < length; i++) {
+					if (i) {
+						putc(' ', output);
+					}
+					fprintf(output, isupper((unsigned char)command[0]) ? "%02X" : "%02x", getc(new_rom));
+				}
+			}
```

Comment:
```suggestion
			}
			free(searchend);
```

Reply:
Resolved in https://github.com/pret/pokecrystal/pull/882/commits/ffaea69f9bd58155927a0dd2a121e08e3df29e73